src/kv/idb-store.ts

```diff
@@ -0,0 +1,155 @@
+import {deleteSentinel, WriteImplBase} from './write-impl-base.js';
+import type {Read, Store, Write} from './store.js';
+
+const RELAXED = {durability: 'relaxed'};
+const OBJECT_STORE = 'chunks';
+
+const enum WriteState {
+  OPEN,
+  COMMITTED,
+  ABORTED,
+}
+
+export class IDBStore implements Store {
+  private readonly _db: Promise<IDBDatabase>;
+
+  constructor(name: string) {
+    this._db = openDatabase(name);
+  }
+
+  async read(): Promise<Read> {
+    const db = await this._db;
+    const tx = db.transaction(OBJECT_STORE, 'readonly');
+    return new ReadImpl(tx);
+  }
+
+  async write(): Promise<Write> {
+    const db = await this._db;
+    // TS does not have type defs for the third options argument yet.
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore Expected 1-2 arguments, but got 3.ts(2554)
+    const tx = db.transaction(OBJECT_STORE, 'readwrite', RELAXED);
+    return new WriteImpl(tx);
+  }
+
+  async close(): Promise<void> {
+    (await this._db).close();
+  }
+}
+
+class ReadImpl {
+  private readonly _tx: IDBTransaction;
+
+  constructor(tx: IDBTransaction) {
+    this._tx = tx;
+  }
+
+  async has(key: string): Promise<boolean> {
+    return (await wrap(objectStore(this._tx).count(key))) > 0;
+  }
+
+  async get(key: string): Promise<Uint8Array | undefined> {
+    return wrap(objectStore(this._tx).get(key));
+  }
+
+  release(): void {
+    // Do nothing. We rely on IDB locking.
+  }
+}
+
+class WriteImpl extends WriteImplBase {
+  private readonly _tx: IDBTransaction;
+  private _txState: Promise<WriteState> | undefined = undefined;
+
+  constructor(tx: IDBTransaction) {
+    super(new ReadImpl(tx));
+    this._tx = tx;
+  }
+
+  private _registerTransaction(): void {
```

Comment:
Couple things confusing to me about how transaction state is tracked:

1. `_registerTransaction` seems misnamed because it doesn't register the transaction with any registry anywhere (I think maybe the Rust code did). All it does is set up some callbacks.
2. There is `_txState` but it's not really a "state", it's a promise that resolves when the transaction closes with its final state. It's really the "end state" of the transaction.
3.  There's also `_transactionState()` which seems related to `_txState` but is different -- it's the current state of the transaction.

What if instead there was `_txState` which wasn't a promise but a normal value of type `WriteState` which reflect the actual state of the tx at this moment. There could also be a `_txStateChange: Promise<void>` that resolves whenever the state changes. At construction time setup the callbacks (currently in `_registerTransaction`) to (a) set the `_txState` correctly and (b) resolve the `_txStateChange` promise and create a new one.

Reply:
Yeah, this is all legacy. Could use cleanup